main.cpp

```diff
@@ -1,15 +1,161 @@
+// #include <cstdio>
+// #include <stdexcept>
 
 #include "helpers.h"
 
-using namespace std;
+// using namespace std;
 
 // main file
-int main(int argc, char **argv)
-{
-    // Code to be written below this comment
+int main(int argc, char **argv) {
+    try {
+        // Define allowable filters
+        const char *filtersOptions {"begrs"};
 
+        // Get filter flag
+        char filter {static_cast<char>(getopt(argc, argv, filtersOptions))};
+
+        // check validity
+        if (filter == '?') {
+            throw std::string {"Invalid filter."};
+        }
+
+        // Ensure only one filter
+        if (getopt(argc, argv, filtersOptions) != -1) {
+            throw std::string {"Only one filter allowed."};
+        }
+
+        // Ensure proper usage
+        if (argc != optind + 2) {
+            throw std::string {"Usage: filter [flag] infile outfile"};
+        }
+
+        // Open input file 
+        FILE *inFile {fopen(argv[optind], "r")};
+        // check readability
+        if (inFile == nullptr) {
+            throw std::string {"Could not read : " + static_cast<std::string>(argv[optind])};
+        }
+
+        // open output file
+        FILE *outFile {fopen(argv[optind + 1], "w")};
+        // check writability
+        if (outFile == nullptr) {
+            throw std::string {"Could not create : " + static_cast<std::string>(argv[optind + 1])};
+        }
+
+        // Read inFile's BITMAPFILEHEADER
+        BITMAPFILEHEADER bf;
+        fread(&bf, sizeof(BITMAPFILEHEADER), 1, inFile);
+
+        // Read inFile's BITMAPINFOHEADER
+        BITMAPINFOHEADER bi;
+        fread(&bi, sizeof(BITMAPINFOHEADER), 1, inFile);
+
+        // Ensure infile is (likely) a 24-bit uncompressed BMP 4.0
+        if (bf.bfType != 0x4d42 || bf.bfoffBits != 54 || bi.biSize != 40 || bi.biBitCount != 24 || bi.biCompression != 0) {
+            fclose(outFile);
+            fclose(inFile);
+            
+            throw std::string {"Unsupported file format."};
+        }
+
+        int height {abs(bi.biHeight)};
+        int width {abs(bi.biWidth)};
+
+        // Allocate memory for image 2D array
+        RGBTRIPLE **image {new RGBTRIPLE *[height]};
+        for (int i{0}; i < height; ++i) {
+               image[i] = new RGBTRIPLE[width];
+        }
+
+        // checks for successful memory allocation
+        if (image == nullptr) {
+            fclose(inFile);
+            fclose(outFile);
+
+            throw std::string {"Not enough memory to store image."};
+        }
+
+        // Determine padding for scanlines
+        size_t padding {(4 - (width * sizeof(RGBTRIPLE)) % 4) % 4};
+
+        // Iterate over infile's scanlines
+        for (int i {0}; i < height; i++) {
+            // Read row into pixel array
+            fread(image + i, sizeof(RGBTRIPLE), width, inFile);
+
+            // Skip over padding
+            fseek(inFile, padding, SEEK_CUR);
+        }
+
+        // Filter image
+        switch (filter)
+        {
```

Comment:
It is advised not to follow this bracket styling.
Please use the bracket styling you used above.